ADD Proprietary Macro Changes from lok repo [MACRO-2201]

synoet opened the pull request:
since `git am -3` is horrible at resolving any kind of conflict / rejection, I had to manually copy paste all the changes over from each commit diff during each patch apply. So its possible some things are missing / got improperly ported over, though I'll do my best to catch anything during testing.

Also some of the commits arent true to their original content since things had to be slightly moved around, but the overall change set is mostly the same.

this excludes commits for:
- v8 related stuff
- .as api
- backups
- colorizer / overlays
- windows / mac specific stuff

**:warning: IMPORTANT: DONT SQUASH THIS PR**

Commit:
Prevent hard crash on unknown LOK enum

Commit:
Expose LOK XComponent and XComponentContext

Commit:
Fix race condition on DoIdleJobs when a SolarMutexGuard is called in LOK

Commit:
Fix layout idling before render, wait for LOK init

chase committed:
Make all boundaries invisible by default in Writer (#22)

synoet committed:
Remove keybindings for quitting + closing windows (#26)

removed all commands related to .uno:Quit, mod+q, mod+shift+q, mod+w

Commit:
Do not notify events for documents without views

whutchinson98 committed:
fix: getCommandValues(.uno:GetOutline)

synoet committed:
ADD setup for page margins

whutchinson98 committed:
fix: Intercept save as command to ensure we use the correct view

synoet committed:
ADD getPageMargins

synoet committed:
ADD getPageSize

synoet committed:
ADD setPageSize

synoet committed:
ADD getPageOrientation

synoet committed:
ADD setPageColor

Commit:
Use postUnoCommand/getCommandValues, fix setPageColor

synoet committed:
FIX apply attr changes on all parts of doc

synoet committed:
CHG desc length

whutchinson98 committed:
MACRO-1266 Do not create Redlines for Insert/Delete comments

whutchinson98 committed:
feat: Create and expose setAuthor call to allow changes to the author after the document is initialized for rendering

synoet committed:
FIX hiding whitespace unrenders document

synoet committed:
ADD link to notion ticket

synoet committed:
ADD link to notion ticket

chase committed:
Remove useless redline warning

Commit:
Add .uno:BatchTrackChange

Commit:
Use .uno:CurrentTrackedChangeId for cursor change

Commit:
MACRO-1384: Fix load document crash

Commit:
MACRO-1384: Fix load document crash

Commit:
MACRO-1392: .uno:UpdateRedlines

Commit:
Simplify redline descriptions

gbirman committed:
Add red/blue colorization and comparison list formatting (#63)

- Adds redlining changes such that strikethroughs are colored red and
additions are blue
- Formatting differences are left unchanged
- Update numbering to match falcon lok comparison number list formatting
before: 2.2[2.1] after:(2.1→2.2)

***Examples***

Consolidate comparison with Christy NDA:
Before:
![CleanShot 2023-09-18 at 17 34
59](https://github.com/coparse-inc/libreofficekit/assets/25272206/56c43e9e-333b-4e70-b1b6-b64dc4ed82b7)
After:
![CleanShot 2023-09-18 at 18 18
26](https://github.com/coparse-inc/libreofficekit/assets/25272206/be26bedf-5d3f-4517-bdf1-0e253d749df6)

Regular `A_Agreement.docx` <--> `B_Agreement.docx` 1:1 comparison:
![CleanShot 2023-09-18 at 18 19
18](https://github.com/coparse-inc/libreofficekit/assets/25272206/e4a2ec9e-a48e-43b4-83a2-ba341d9527d2)

Formatting differences are left unchanged (except for additions which
are blue):
![CleanShot 2023-09-18 at 18 22
19](https://github.com/coparse-inc/libreofficekit/assets/25272206/3b0d81e3-c46a-4518-8a80-9a41fc18c34c)

synoet committed:
FIX disable LanguageGuesser

mikekaganski committed:
Do not forget to notify in cases when modified redline's start/end don't move

SwRangeRedline::SetStart/SetEnd call MaybeNotifyRedlineModification,
which means that e.g. continuing typing at the end of the current
redline would continue sending notifications. But as soon as cursor
is moved (e.g. using arrow left, inside the current redline), the
next modification is completely inside the current one, and would
be simply deleted. In this case (and other similar), notifications
weren't sent.

This change tries to make sure that every such case is handled. In
case of delete-inside-insert, when positions are equal, the current
redline is deleted; MaybeNotifyRedlineModification is not called,
to not crash. I don't know if this needs additional notification.

Change-Id: I2a97354623a5a784928ce4e1f2c93606954bd11c
Reviewed-on: https://gerrit.libreoffice.org/c/core/+/158198
Tested-by: Jenkins
Reviewed-by: Mike Kaganski <mike.kaganski@collabora.com>
(cherry picked from commit 34ac12dca3f5af50fddfb7c77e2943897980b815)
Reviewed-on: https://gerrit.libreoffice.org/c/core/+/158163
Tested-by: Jenkins CollaboraOffice <jenkinscollaboraoffice@gmail.com>
Reviewed-by: Miklos Vajna <vmiklos@collabora.com>

Commit:
Encourage invalidaton after batch update of track change

gbirman committed:
No bold in LOK changed attribute

Commit:
Remove dead theme code that causes performance issues

Commit:
Make saving large documents with a large amount of bookmarks 11x faster

Commit:
Make saving large documents with a large amount of bookmarks 11x faster

Commit:
Prevent crash caused by freeing non-trivial destructor of the static ImplSVData instance

Commit:
Do not encode empty clipboard strings as HTML

Commit:
Fix missing clipboard listener cases

rithy58 committed:
Do not mutate text when reference is not found

synoet committed:
CHG ignore read only flag

synoet committed:
FIX

synoet committed:
MV comment

gbirman commented:
Squashing time

synoet commented:
I can prob rebase and drop my self as the co-author from all these commits, though realistically each one has some small changes

chase commented:
> I can prob rebase and drop my self as the co-author from all these commits, though realistically each one has some small changes

Don't bother.

Also, we can drop the code for lokdocumenteventnotifier and related, the document event hooks caused problems @synoet

This still needs to be reviewed, unnecessary code needs to be eliminated.

I'm going to mark this as draft until this is thoroughly tested and cleaned up.

chase commented on VERSIOn:
Remove

chase commented on libreoffice-core/Repository.mk:
```diff
@@ -1040,6 +1040,7 @@ $(eval $(call gb_Helper_register_packages_for_install,ooo,\
 		helpcontent2_html_static \
 	) \
 	resource_fonts \
+	custom_macro_fonts \
```
Remove, not necessary

chase commented on libreoffice-core/comphelper/Library_comphelper.mk:
```diff
@@ -168,6 +168,7 @@ $(eval $(call gb_Library_add_exception_objects,comphelper,\
     comphelper/source/xml/attributelist \
     comphelper/source/xml/ofopxmlhelper \
     comphelper/source/xml/xmltools \
+    comphelper/source/misc/inthash \
```
Remove, not necessary

synoet replied:
@chase Are you sure its used in `libreoffice-core/sw/source/filter/ww8/wrtww8.hxx`, for `MACRO-1786`

chase commented on libreoffice-core/desktop/Library_sofficeapp.mk:
```diff
@@ -130,6 +130,7 @@ $(eval $(call gb_Library_add_exception_objects,sofficeapp,\
 ifneq ($(filter $(OS),ANDROID iOS MACOSX WNT),)
 $(eval $(call gb_Library_add_exception_objects,sofficeapp,\
 	desktop/source/lib/init \
+    desktop/source/lib/lokdocumenteventnotifier \
```
Remove, not necessary

chase commented on libreoffice-core/desktop/inc/lib/init.hxx:
```diff
@@ -250,6 +251,7 @@ namespace desktop {
     {
         css::uno::Reference<css::lang::XComponent> mxComponent;
         std::shared_ptr< LibreOfficeKitDocumentClass > m_pDocumentClass;
+        std::unique_ptr<doceventnotifier::LokDocumentEventNotifier> m_lokDocEventNotifier;
```
Revert the commit that introduced this, not necessary

chase commented on libreoffice-core/desktop/inc/lib/init.hxx:
```diff
@@ -34,6 +34,7 @@
 #include <sfx2/lokhelper.hxx>
 
 #include <desktop/dllapi.h>
+#include "../source/lib/lokdocumenteventnotifier.hxx"
```
Revert the commit that introduced this, not necessary

chase commented on libreoffice-core/desktop/source/lib/init.cxx:
```diff
@@ -3003,6 +3137,18 @@ static LibreOfficeKitDocument* lo_documentLoadWithOptions(LibreOfficeKit* pThis,
             pDocument->maFontsMissing.insert(aFontMappingUseData[i].mOriginalFont);
         }
 
+        // [MACRO-2259] We don't respect the readonly flag.
+        // If the document is loaded readonly, we need to set it to false.
```
Long ranges (not a couple lines) of Macro changes should be paired with an opening and closing like:
```
// MACRO: We are doing weird stuff {
...
...
...
...
// MACRO: }
```

This makes it easier to cut out or identify our changes to upstream during merge conflicts _and_ if need be, remove it if/when we requested to open source things.

chase commented on libreoffice-core/desktop/source/lib/init.cxx:
```diff
@@ -907,6 +911,127 @@ void ExecuteMarginULChange(
                                                           SfxCallMode::RECORD, { pPageULMarginItem });
 }
 
+void setPageColor(LibreOfficeKitDocument* pThis, const OUString rColorHex)
```
See comment above about `// MACRO: ... {` comments.

Any proprietary change should be wrapped in it, excluding `#include` statements at the top.

chase commented on libreoffice-core/desktop/source/lib/init.cxx:
```diff
@@ -1144,6 +1269,8 @@ static void doc_getDataArea(LibreOfficeKitDocument* pThis,
 static void doc_initializeForRendering(LibreOfficeKitDocument* pThis,
                                        const char* pArguments);
 
+static void doc_setAuthor(LibreOfficeKitDocument* pThis, const char* sAuthor);
```
For single line proprietary changes, a single `/// MACRO: ` line is sufficient. If there is no ticket ID explicitly tied to it or it's hard to find the ticket ID, it can be omitted.

chase commented on libreoffice-core/desktop/source/lib/init.cxx:
```diff
@@ -4613,6 +4782,8 @@ static void doc_registerCallback(LibreOfficeKitDocument* pThis,
     if (pCallback != nullptr)
     {
+        uno::Reference<css::lang::XComponent> mxOwner(pDocument->mxComponent, uno::UNO_QUERY);
+
```
This seems extraneous. mxOwner is unused.

chase commented on libreoffice-core/desktop/source/lib/init.cxx:
```diff
@@ -5499,6 +5835,105 @@ static bool getFromTransferable(
     const css::uno::Reference<css::datatransfer::XTransferable> &xTransferable,
     const OString &aInMimeType, OString &aRet);
 
+namespace {
+    char* leakyStrDup(const std::string_view& view) {
+        char* pMemory = static_cast<char*>(malloc(view.size() + 1));
+        assert(pMemory); // don't tolerate failed allocations.
+        std::memcpy(pMemory, view.data(), view.size());
+        pMemory[view.size()] = '\0';
+        return pMemory;
+    }
+}
```
New ticket and PR (@synoet please write it up for me and start on it as a medium priority after this is done):
We should rework any code that uses this, move them to wasm_extensions.hxx/wasm_extensions.cxx to return `std::string` and let embind handle the string lifetime appropriately.

Leaking memory isn't something we can afford since memory is limited to 2GB and each page that's allocated is 64KiB.

synoet replied:
https://www.notion.so/macrocom/Move-things-that-use-leakyStrDup-to-wasm-extensions-eebd0795686347e28f653587ca6b2ddf

chase commented on libreoffice-core/desktop/source/lib/lokdocumenteventnotifier.cxx:
PR for this should be reverted, this code removed

chase commented on libreoffice-core/desktop/source/lib/lokdocumenteventnotifier.hxx:
PR for this should be reverted, this code removed

chase commented on libreoffice-core/extras/Module_extras.mk:
```diff
@@ -36,6 +36,7 @@ $(eval $(call gb_Module_add_targets,extras,\
 	Package_palettes \
 	Package_tplwizdesktop \
 	Package_wordbook \
+	Package_custom_macro_fonts \
```
PR for this should be reverted, this code removed

chase commented on libreoffice-core/extras/Package_custom_macro_fonts.mk:
PR for this should be reverted, this code removed

chase commented on libreoffice-core/extras/source/macro_fonts/Caladea-Bold.ttf:
PR for this should be reverted, macro_fonts should be removed

chase commented on libreoffice-core/include/LibreOfficeKit/LibreOfficeKitEnums.h:
PR for these changes should be reverted, these changes removed

chase commented on libreoffice-core/include/LibreOfficeKit/LibreOfficeKitInit.h:
This change isn't accurate anymore, should be reverted

chase commented on libreoffice-core/include/rtl/string.hxx:
These changes can be reverted, we compile with C++ exception support since WASM supports them natively now.

Any code added with a new condition of `#if defined SAL_NO_EXCEPT` can safely be reverted.

chase commented on libreoffice-core/include/rtl/ustring.hxx:
See above comment. Can be reverted.

chase commented on libreoffice-core/officecfg/registry/data/org/openoffice/Office/Common.xcu:
PR for these changes should be reverted, these changes removed

chase commented on libreoffice-core/officecfg/registry/schema/org/openoffice/Office/Writer.xcs:
```diff
@@ -2526,7 +2526,11 @@
             <value>0</value><!-- MACRO: Disable bold on LOK changed attribute (since we disabled coloring) -->
+=======
+            <value>0</value>
+>>>>>>> No bold in LOK changed attribute
```
Unresolved merge conflict. Choose HEAD.
```suggestion
            <value>0</value><!-- MACRO: Disable bold on LOK changed attribute (since we disabled coloring) -->
```

chase commented on libreoffice-core/sfx2/source/doc/objstor.cxx:
```diff
@@ -2720,6 +2720,7 @@ bool SfxObjectShell::CommonSaveAs_Impl(const INetURLObject& aURL, const OUString
         return false;
     }
 
+    // HERE
```
Extraneous change, should be removed

chase commented on libreoffice-core/sw/inc/viewopt.hxx:
```diff
@@ -831,8 +831,8 @@ public:
-    bool IsTableBoundaries() const {return IsAppearanceFlag(ViewOptFlags::TableBoundaries);}
-    bool IsIndexShadings() const {return IsAppearanceFlag(ViewOptFlags::IndexShadings);}
+    bool IsTableBoundaries() const {return false;}
+    bool IsIndexShadings() const {return false;}
```
Since support was added upstream for toggling these, this change should be reverted instead of hard-coded.

chase commented on libreoffice-core/sw/source/core/text/macro_internal.cxx:
This is colorizer code. This PR/these changes should be reverted.

chase commented on libreoffice-core/sw/source/core/text/atrstck.cxx:
This is colorizer code. This PR/these changes should be reverted.

chase commented on libreoffice-core/sw/source/core/text/macro_internal.hxx:
This is colorizer code. This PR/these changes should be reverted.

chase commented on libreoffice-core/sw/source/filter/ww8/docxexport.hxx:
Should probably revert this extraneous change (not the commit, just the changes to this file). It makes it easier to handle merge conflicts later.

chase commented on patches/0.patch:
Looks like you accidentally added a temporary `patches` directory to this PR.

synoet replied:
yeah sorry, I included them temporarily just to keep track of them, when I had to make changes which patches I explicitly dropped. I'll remove them ofcourse before merging

synoet replied:
removed

chase reviewed (changes requested):
See my comments below.

Also please go back and add the `MACRO:` comments I mentioned in a comment below to all of our proprietary code.

synoet committed:
RM unresolved merge conflicts

synoet committed:
ADD comments to all macro prop changes

chase commented:
Oops, forgot to delete that comment. You're right, it's used for the export
optimization on Flutter.

On Thu, Apr 18, 2024, 20:16 teo ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In libreoffice-core/comphelper/Library_comphelper.mk
> <https://github.com/coparse-inc/lok-wasm/pull/18#discussion_r1570728532>:
>
> > @@ -168,6 +168,7 @@ $(eval $(call gb_Library_add_exception_objects,comphelper,\
>      comphelper/source/xml/attributelist \
>      comphelper/source/xml/ofopxmlhelper \
>      comphelper/source/xml/xmltools \
> +    comphelper/source/misc/inthash \
>
> @chase <https://github.com/chase> Are you sure its used in
> libreoffice-core/sw/source/filter/ww8/wrtww8.hxx, for MACRO-1786
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/coparse-inc/lok-wasm/pull/18#discussion_r1570728532>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABKI52O5Z67BIWFGEPHYDY57BTPAVCNFSM6AAAAABGKEBZ6KVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAMBYHA3TCNRVGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

synoet committed:
ADD more comments

synoet committed:
ADD loadFromMemory, saveToMemory

synoet committed:
RM startBatchUpdate, finishBatchUpdate

synoet commented:
@chase Update: 
- removed all unneeded changes and dropped their commits
- Added // MACRO: comments to all of our changes 
- Added back loadDocumentFromMemory , saveToMemory
- things are compiling / working as expected

Only thing i noticed is blue additions are not underlines.

synoet committed:
CHG make insertions underlined

synoet committed:
Merge branch 'main' of github.com:coparse-inc/lok-wasm into teo/add/proprietary-changes

synoet commented:
@chase fixed underlines + merged in main
